### PR TITLE
Updated MarkDuplicates code to rely on the Picard OpticalDuplicatesFinder

### LINKED ISSUE #4700: Update MarkDuplicatesSpark OpticalDuplicate code to match Picards
Picard had its Optical duplicate finding code refactored recently, additionally it has been noticed as part of #4656 that we are currently not properly accounting for the read groups when we stratify reads in MarkDuplicatesSpark which will likely cause problems for bams with more than one read group. Additionally better test coverage for multiple read groups should be added to ensure we are handling them sanely.

### PR DESCRIPTION
**Pull request by jamesemery**
Fixes #4700

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java [hunk 1]
```diff
@@ -86,6 +92,9 @@ public int getIndex() {
 
         final JavaPairRDD<String, Iterable<IndexPair<GATKRead>>> keyedReads = getReadsGroupedByName(header, mappedReads, numReducers);
 
+        //TODO why in the world does a JavaRDD return a normal sparkcontext instead of the nice wrapped one...?
+        final Broadcast<Map<String, Short>> headerReadGroupIndexMap = new JavaSparkContext(reads.context()).broadcast( getHeaderReadGroupIndexMap(header));
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java]: use `JavaSparkContext.fromSparkContext` instead I think.

File: src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/OpticalDuplicateFinder.java [hunk 2]
```diff
@@ -1,243 +0,0 @@
...
-        for (int i = 0; i < readName.length(); i++) {
-            if (readName.charAt(i) == delim) {
-                if (1 < tokensIdx && tokensIdx < 5)
-                    tokens[tokensIdx] = rapidParseInt(readName.substring(prevIdx, i)); // only fill in 2-4 inclusive
-                tokensIdx++;
-                if (4 < tokensIdx) return tokensIdx; // early return, only consider the first five tokens
-                prevIdx = i + 1;
-            }
-        }
-        if (prevIdx < readName.length()) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/OpticalDuplicateFinder.java]: YAY deleting code!

File: src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/ReadEnds.java [hunk 3]
```diff
@@ -36,16 +36,16 @@
...
-    public void setX(final short x) { this.x = x; }
+    public void setX(final int x) { this.x = (short)x; }
 
     @Override
-    public short getY() { return this.y; }
+    public int getY() { return this.y; }
 
     @Override
-    public void setY(final short y) { this.y = y; }
+    public void setY(final int y) { this.y = (short)y; }
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/ReadEnds.java]: We should add checks here that it's not > maxshort.

File: src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/SerializableOpticalDuplicatesFinder.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package org.broadinstitute.hellbender.utils.read.markduplicates;
+
+import picard.sam.markduplicates.util.OpticalDuplicateFinder;
+import java.io.Serializable;
+
+/**
+ * An intermediate class flagged as being serializable so that the Picard OpticalDuplicateFinder can be serialized despite
+ * it not being marked as such. There is not state outside of construction at the time of creating this pr.
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/SerializableOpticalDuplicatesFinder.java]: the second sentence is more confusing than helpful

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/markduplicates/EstimateLibraryComplexityGATK.java [hunk 5]
```diff
@@ -142,11 +142,11 @@
     /**
      * Little class to hold the sequence of a pair of reads and tile location information.
      */
-    static class PairedReadSequence implements OpticalDuplicateFinder.PhysicalLocation {
+    static class PairedReadSequence implements picard.sam.util.PhysicalLocation {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/markduplicates/EstimateLibraryComplexityGATK.java]: do we need the full path here?

File: src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/ReadEnds.java [hunk 6]
```diff
@@ -1,7 +1,7 @@
 package org.broadinstitute.hellbender.utils.read.markduplicates;
 
 /** Little struct-like class to hold read pair (and fragment) end data for duplicate marking. */
-public abstract class ReadEnds implements OpticalDuplicateFinder.PhysicalLocation {
+public abstract class ReadEnds implements picard.sam.util.PhysicalLocation {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/ReadEnds.java]: shorten the path?

File: src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/sparkrecords/Pair.java [hunk 7]
```diff
@@ -20,7 +19,7 @@
  * during the processing step of MarkDuplicatesSpark
  */
 @DefaultSerializer(Pair.Serializer.class)
-public final class Pair extends PairedEnds implements OpticalDuplicateFinder.PhysicalLocation {
+public final class Pair extends PairedEnds implements picard.sam.util.PhysicalLocation {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/sparkrecords/Pair.java]: short path?

File: src/main/java/org/broadinstitute/hellbender/utils/test/testers/SamFileTester.java [hunk 8]
```diff
@@ -225,6 +230,11 @@ public void addMatePair(final String readName,
             record2.setAttribute("MC", null);
         }
 
+        if (readGorup!=null) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/test/testers/SamFileTester.java]: typo: readGorup

image

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/markduplicates/AbstractMarkDuplicatesCommandLineProgramTest.java [hunk 9]
```diff
@@ -170,6 +167,30 @@ public void testOpticalDuplicateClusterSamePositionNoOpticalDuplicates() {
         tester.runTest();
     }
 
+    @Test
+    public void testOpticalDuplicatesDifferentReadGroups() {
+        final AbstractMarkDuplicatesTester tester = getTester();
+        tester.setExpectedOpticalDuplicate(0);
+        tester.addMatePair("RUNID:7:1203:2886:82292",  19, 19, 485253, 485253, false, false, true, true, "42M59S", "59S42M", true, false, false, false, false, DEFAULT_BASE_QUALITY, "H0164.2");  // duplicate
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/markduplicates/AbstractMarkDuplicatesCommandLineProgramTest.java]: I hate these 600 parameter methods... but I'm not sure what to do about them

File: src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java (hunk 1 above)
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/OpticalDuplicateFinder.java (hunk 2 above)
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/OpticalDuplicateFinder.java]: Huzzah!

File: src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/ReadEnds.java (hunk 3 above)
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/ReadEnds.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/SerializableOpticalDuplicatesFinder.java (hunk 4 above)
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/SerializableOpticalDuplicatesFinder.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/testers/SamFileTester.java (hunk 8 above)
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/utils/test/testers/SamFileTester.java]: done

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/markduplicates/AbstractMarkDuplicatesCommandLineProgramTest.java (hunk 9 above)
Comment (jamesemery) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/markduplicates/AbstractMarkDuplicatesCommandLineProgramTest.java]: there are a bunch of overloads that help the problem. For the most par every field is necessary to at least one of these tests.

**Comment by jamesemery** [comment]
@lbergelson Responded to your comments.

**Comment by jamesemery** [comment]
@lbergelson Is this branch good to go?

**Comment by lbergelson** [comment]
@jamesemery We have test failures again https://storage.googleapis.com/hellbender-test-logs/build_reports/master_19079.2/tests/test/index.html

**Comment by jamesemery** [comment]
@lbergelson Yeah, so the test was broken because before it the pairedEnds y value was getting truncated. Now that test is running on two reads that are ACTUALLY close enough for comparison.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java [hunk 10]
```diff
@@ -133,14 +133,17 @@ public String toString() {
...
             // If there are two primary reads in the group pass them as a pair
             } else if (primaryReads.size()==2) {
-                final IndexPair<GATKRead> firstRead = primaryReads.get(0);
+                final GATKRead firstRead = primaryReads.get(0).getValue();
                 final IndexPair<GATKRead> secondRead = primaryReads.get(1);
-                final Pair pair = MarkDuplicatesSparkRecord.newPair(firstRead.getValue(), secondRead.getValue(), header, secondRead.getIndex(), scoringStrategy);
-                final Short readGroup = headerReadGroupIndexMap.getValue().get(firstRead.getValue().getReadGroup());
+                final Pair pair = MarkDuplicatesSparkRecord.newPair(firstRead, secondRead.getValue(), header, secondRead.getIndex(), scoringStrategy);
+                // Validate and add the read group to the pair
+                final Short readGroup = headerReadGroupIndexMap.getValue().get(firstRead.getReadGroup());
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java]: could we add 2 simple tests that show that it explodes with the right errors on each of these cases

**Comment by lbergelson** [comment]
https://storage.googleapis.com/hellbender-test-logs/build_reports/master_19134.2/tests/test/index.html

**Pull request merged**